### PR TITLE
[Seq] Replace uninitialized aggregate registers with constant zeros

### PR DESCRIPTION
**Pull request by uenoku**
This PR implements a canonicalization to replace uninitialized aggregate registers with constant zeros. 
This fixes lint warnings with aggregate preservation builds. Eventually the canonicalization should be moved to a folder once aggregate constant is added.

### CODE REVIEW

**Comment by dtzSiFive** [comment]
Shouldn't this consider reset signal/value? If it's feedback, replace with reset value, etc....  I think if clock is constant an async reset can still reach the register... ?

The spec says behavior of constructs with indeterminate values is implementation-defined but constrained to initialize registers consistently, which seems to suggest we can't replace with zero only with the global initialization value/pattern  (sometimes random)?
That may be intended to only apply to registers that aren't optimized away (by logic such as in this PR) when not considering such a constraint? Do you know?

**Comment by uenoku** [comment]
> Shouldn't this consider reset signal/value? If it's feedback, replace with reset value, etc.... I think if clock is constant an async reset can still reach the register... ?

Good point, basically this is just a port of folder to handle aggregates (currently we cannot handle aggregate in a folder because we don't have constant integers).

**Review by youngar — APPROVED** [review]
Looks good.  I've never seen the `@Observe` pattern before used for the other tests, it seems like we can't leverage it since it only takes an `i32`. I don't think we can use type parameters in `hw` modules.  We could also replace it with an `sv.printf` or any unregistered op. @fabianschuiki any idea?

File: lib/Dialect/Seq/SeqOps.cpp [hunk 1]
```diff
@@ -288,6 +288,28 @@ LogicalResult FirRegOp::canonicalize(FirRegOp op, PatternRewriter &rewriter) {
...
+    // If the register has a reset value, we can replace it with that.
+    if (auto resetValue = op.getResetValue()) {
+      rewriter.replaceOp(op, resetValue);
+      return success();
+    }
+
+    auto constant = rewriter.create<hw::ConstantOp>(
+        op.getLoc(), APInt::getZero(hw::getBitWidth(op.getType())));
+    rewriter.replaceOpWithNewOp<hw::BitcastOp>(op, op.getType(), constant);
+    return success();
```
Comment (fabianschuiki) [inline lib/Dialect/Seq/SeqOps.cpp:310]: Oh using a bitcast here is pretty clever!

**Review by fabianschuiki — APPROVED** [review]
This looks like a good addition!

**Comment by fabianschuiki** [comment]
> Looks good. I've never seen the `@Observe` pattern before used for the other tests, it seems like we can't leverage it since it only takes an `i32`. I don't think we can use type parameters in `hw` modules. We could also replace it with an `sv.printf` or any unregistered op. @fabianschuiki any idea?

Hmmm yeah I just hacked that in there to get a black hole use of the value that is easy to identify and observe. A print would totally also work! Or an unrealized conversion cast that casts to no value at all. You could also create a `@Observe2` that takes the right kind of argument 😬

**Comment by uenoku** [comment]
I think `FirRegAggregate` test is small so we can just use `hw.output` for that purpose :+1:

**Pull request merged**